File: javascript/src/pools.js
const { ethers } = require('ethers');
const fs = require('fs');
const path = require('path');
const cliProgress = require('cli-progress');

const { logger, CACHED_POOLS_FILE } = require('./constants');

const Erc20Abi = [
    'function decimals() external view returns (uint8)'
];

const V2FactoryAbi = [
    'event PairCreated(address indexed token0, address indexed token1, address pair, uint)'
];

const DexVariant = {
    UniswapV2: 2,
    UniswapV3: 3,
};

class Pool {
    constructor(
        address,
        version,
        token0,
        token1,
        decimals0,
        decimals1,
        fee
    ) {
        this.address = address;
        this.version = version;
        this.token0 = token0;
        this.token1 = token1;
        this.decimals0 = decimals0;
        this.decimals1 = decimals1;
        this.fee = fee;
    }

    cacheRow() {
        return [
            this.address,
            this.version,
            this.token0,
            this.token1,
            this.decimals0,
            this.decimals1,
            this.fee,
        ];
    }
}

const range = (start, stop, step) => {
    let loopCnt = Math.ceil((stop - start) / step);
    let rangeArray = [];
    for (let i = 0; i < loopCnt; i++) {
        let fromBlock = start + (i * step);
        let toBlock = Math.min(fromBlock + step, stop);
        rangeArray.push([fromBlock, toBlock]);
    }
    return rangeArray;
}

function loadCachedPools() {
    let cacheFile = path.join(__dirname, '..', CACHED_POOLS_FILE);
    let pools = {}
    if (fs.existsSync(cacheFile)) {
        const content = fs.readFileSync(cacheFile, 'utf-8');
        const rows = content.split('\n');
        for (let row of rows) {
            if (row == '') continue;
            row = row.split(',');
            if (row[0] == 'address') continue;
            let version = row[1] == '2' ? DexVariant.UniswapV2 : DexVariant.UniswapV3;
            let pool = new Pool(row[0],
                                version,
                                row[2],
                                row[3],
                                parseInt(row[4]),
                                parseInt(row[5]),
                                parseInt(row[6]))
            pools[row[0]] = pool;
        }
    }
    return pools;
}

function cacheSyncedPools(pools) {
    const columns = ['address', 'version', 'token0', 'token1', 'decimals0', 'decimals1', 'fee'];
    let data = columns.join(',') + '\n';
    for (let address in pools) {
        let pool = pools[address];
        let row = pool.cacheRow().join(',') + '\n';
        data += row;
    }
    let cacheFile = path.join(__dirname, '..', CACHED_POOLS_FILE);
    fs.writeFileSync(cacheFile, data, { encoding: 'utf-8' });
}

async function loadAllPoolsFromV2(
    httpsUrl,
    factoryAddresses,
    fromBlocks,
    chunk
) {
    /*
    Retrieves historical events from Uniswap V2 factories.

    Whenever a new pool is created from the Uniswap V2 factory,
    a "PairCreated" event is emitted. We request for all the PairCreated
    events from the block these factories were deployed.

    👉 NOTE: the process takes a really long time, because it has room for improvement.
    This function will make requests to the RPC endpoint one batch at a time,
    each looking at events from block range of: [fromBlock, toBlock] chunk size.
    */
    let pools = loadCachedPools();
    if (Object.keys(pools).length > 0) {
        return pools;
    }

    const provider = new ethers.providers.JsonRpcProvider(httpsUrl);
    const toBlock = await provider.getBlockNumber();
    
    const decimals = {};
    pools = {};

    for (let i = 0; i < factoryAddresses.length; i++) {
        const factoryAddress = factoryAddresses[i];
        const fromBlock = fromBlocks[i];

        const v2Factory = new ethers.Contract(factoryAddress, V2FactoryAbi, provider);

        const requestParams = range(fromBlock, toBlock, chunk);

        const progress = new cliProgress.SingleBar({}, cliProgress.Presets.shades_classic);
        progress.start(requestParams.length);

        for (let i = 0; i < requestParams.length; i++) {
            const params = requestParams[i];
            const filter = v2Factory.filters.PairCreated;
            const events = await v2Factory.queryFilter(filter, params[0], params[1]);

            for (let event of events) {
                let token0 = event.args[0];
                let token1 = event.args[1];

                let decimals0;
                let decimals1;

                try {
                    if (token0 in decimals) {
                        decimals0 = decimals[token0];
                    } else {
                        let token0Contract = new ethers.Contract(token0, Erc20Abi, provider);
                        decimals0 = await token0Contract.decimals();
                        decimals[token0] = decimals0;
                    }

                    if (token1 in decimals) {
                        decimals1 = decimals[token1];
                    } else {
                        let token1Contract = new ethers.Contract(token1, Erc20Abi, provider);
                        decimals1 = await token1Contract.decimals();
                        decimals[token1] = decimals1;
                    }
                } catch (_) {
                    // some token contracts don't exist anymore: eth_call error
                    logger.warn(`Check if tokens: ${token0} / ${token1} still exists`);
                    continue;
                }

                let pool = new Pool(event.args[2],
                                    DexVariant.UniswapV2,
                                    token0,
                                    token1,
                                    decimals0,
                                    decimals1,
                                    300);
                pools[event.args[2]] = pool;
            }

            progress.update(i + 1);
        }

        progress.stop();
    }

    cacheSyncedPools(pools);
    return pools;
}

module.exports = {
    loadAllPoolsFromV2,
};

File: python/pools.py
import os
import csv
import web3
import json

from tqdm import tqdm
from enum import Enum
from web3 import Web3
from time import sleep
from random import randint
from typing import Dict, List, Optional
from concurrent.futures import ThreadPoolExecutor, as_completed

from constants import *  

RATE_LIMIT = 10  # Limit requests, adjust this to rate limit 


class DexVariant(Enum):
    UniswapV2 = 2
    UniswapV3 = 3


class Pool:

    def __init__(self,
                 address: str,
                 version: DexVariant,
                 token0: str,
                 token1: str,
                 decimals0: int,
                 decimals1: int,
                 fee: int):

        self.address = address
        self.version = version
        self.token0 = token0
        self.token1 = token1
        self.decimals0 = decimals0
        self.decimals1 = decimals1
        self.fee = fee

    def cache_row(self):
        return [
            self.address,
            self.version.value,
            self.token0,
            self.token1,
            self.decimals0,
            self.decimals1,
            self.fee,
        ]
        

def fetch_events(params: tuple,
                 factory_address: str,
                 v2_factory: web3.contract.Contract):
    sleep(randint(10, 50) / 100.0)  # Adding some jitter to avoid rate-limiting
    try:
        events = v2_factory.events.PairCreated.get_logs(fromBlock=params[0], toBlock=params[1])
        return [(factory_address, event) for event in events]
    except Exception as e:
        print(f'Error fetching events: {e}')
        return []


def load_cached_pools() -> Optional[Dict[str, Pool]]:
    if os.path.exists(CACHED_POOLS_FILE):
        f = open(CACHED_POOLS_FILE, 'r')
        rdr = csv.reader(f)
        pools = {}

        for row in rdr:
            if row[0] == 'address':
                continue
            version = DexVariant.UniswapV2 if row[1] == '2' else DexVariant.UniswapV3
            pool = Pool(address=row[0],
                        version=version,
                        token0=row[2],
                        token1=row[3],
                        decimals0=int(row[4]),
                        decimals1=int(row[5]),
                        fee=int(row[6]))
            pools[row[0]] = pool
        logger.info(f'Loaded pools from cache: {CACHED_POOLS_FILE} ({len(pools)} pools)')

        return pools


def cache_synced_pools(pool: Pool):
    if os.path.exists(CACHED_POOLS_FILE):
        f = open(CACHED_POOLS_FILE, 'r')
        rdr = csv.reader(f)
        existing_pools = [row[0] for row in rdr]
        f.close()
        if pool.address in existing_pools:
            return
        f = open(CACHED_POOLS_FILE, 'a', newline='')
    else:
        f = open(CACHED_POOLS_FILE, 'w', newline='')
        wr = csv.writer(f)
        columns = ['address', 'version', 'token0', 'token1', 'decimals0', 'decimals1', 'fee']
        wr.writerow(columns)

    wr = csv.writer(f)  
    wr.writerow(pool.cache_row())
    f.close()


def load_all_pools_from_v2(https_url: str,
                           factory_addresses: List[str],
                           from_blocks: List[int],
                           chunk: int = 100000) -> Dict[str, Pool]:

    # Load cached pools
    pools = load_cached_pools()
    
    # If cached pools exist, skip the fetching part
    if pools is not None:
        print("Pools already exist in cache. Skipping fetching.")
        return pools
    
    pools = pools or {}
    v2_factory_abi = json.load(open(ABI_PATH / 'UniswapV2Factory.json', 'r'))
    erc20_abi = json.load(open(ABI_PATH / 'ERC20.json', 'r'))
    w3 = Web3(Web3.HTTPProvider(https_url))
    to_block = w3.eth.get_block_number()
    decimals: Dict[str, int] = {}

    with ThreadPoolExecutor(max_workers=RATE_LIMIT) as executor:
        rate_limit_futures = []

        for i in range(len(factory_addresses)):
            factory_address = factory_addresses[i]
            from_block = from_blocks[i]
            v2_factory = w3.eth.contract(address=factory_address, abi=v2_factory_abi)

            block_range = list(range(from_block, to_block, chunk))
            request_params = [(block_range[i], block_range[i + 1]) for i in range(len(block_range) - 1)]

            for params in request_params:
                future = executor.submit(fetch_events, params, factory_address, v2_factory)
                rate_limit_futures.append(future)

        with tqdm(total=len(rate_limit_futures), desc='Processing events', ascii=' =', leave=True) as pbar:
            for future in as_completed(rate_limit_futures):
                for factory_address, event in future.result():
                    args = event.args
                    token0 = args.token0
                    token1 = args.token1

                    try:
                        if token0 in decimals:
                            decimals0 = decimals[token0]
                        else:
                            token0_contract = w3.eth.contract(address=token0, abi=erc20_abi)
                            decimals0 = token0_contract.functions.decimals().call()
                            decimals[token0] = decimals0

                        if token1 in decimals:
                            decimals1 = decimals[token1]
                        else:
                            token1_contract = w3.eth.contract(address=token1, abi=erc20_abi)
                            decimals1 = token1_contract.functions.decimals().call()
                            decimals[token1] = decimals1
                    except Exception as _:
                        continue

                    pool = Pool(address=args.pair,
                                version=DexVariant.UniswapV2,
                                token0=args.token0,
                                token1=args.token1,
                                decimals0=decimals0,
                                decimals1=decimals1,
                                fee=300)
                    if args.pair not in pools:
                        pools[args.pair] = pool
                        cache_synced_pools(pool)

                pbar.update(1)

    return pools


if __name__ == '__main__':
    """
    Example code can be run on Ethereum
    """
    factory_addresses = [
        '0xC0AEe478e3658e2610c5F7A4A2E1777cE9e4f2Ac', # Uniswap v2
    ]
    factory_blocks = [
        10794229,
    ]

    pools = load_all_pools_from_v2(HTTPS_URL, factory_addresses, factory_blocks, 100000)
    print(pools)